src/frontend/flight/sync/schema.rs

```diff
@@ -0,0 +1,152 @@
+use crate::frontend::flight::sync::SyncError;
+use arrow_schema::{DataType, FieldRef, SchemaRef};
+use clade::sync::{ColumnDescriptor, ColumnRole};
+use std::collections::HashSet;
+
+#[derive(Clone, Debug)]
+pub struct SyncSchema {
+    schema: SchemaRef,
+    columns: Vec<SyncColumn>,
+}
+
+impl SyncSchema {
+    pub fn try_new(
+        column_descriptors: Vec<ColumnDescriptor>,
+        schema: SchemaRef,
+    ) -> Result<Self, SyncError> {
+        if column_descriptors.len() != schema.all_fields().len() {
+            return Err(SyncError::SchemaError {
+                reason: "Column descriptors do not match the schema".to_string(),
+            });
+        }
+
+        // Validate field role's are parsable, we have the correct number of old/new PKs,
+        // and Changed role is non-nullable boolean type which points to an existing column
+        // TODO: Validate a column can not be a PK and Value at the same time
+        let mut old_pk_types = HashSet::new();
+        let mut new_pk_types = HashSet::new();
+        for (col_desc, field) in column_descriptors.iter().zip(schema.fields()) {
+            match ColumnRole::try_from(col_desc.role) {
+                Ok(ColumnRole::OldPk) => {
+                    old_pk_types.insert((&col_desc.name, field.data_type().clone()));
+                }
+                Ok(ColumnRole::NewPk) => {
+                    new_pk_types.insert((&col_desc.name, field.data_type().clone()));
+                }
+                Ok(ColumnRole::Value) => {}
+                Ok(ColumnRole::Changed) => {
+                    let err = if field.data_type() != &DataType::Boolean {
+                        format!(
+                            "Field for column with `Changed` role must be of type boolean: {}",
+                            field.name()
+                        )
+                    } else if field.is_nullable() {
+                        format!(
+                            "Field for column with `Changed` role can not be nullable: {}",
+                            field.name()
+                        )
+                    } else if !column_descriptors.iter().any(|other_cd| {
+                        col_desc.name == other_cd.name
+                            && other_cd.role == ColumnRole::Value as i32
```

Comment:
Is this going to be robust to us adding new members / changing the grpc enum? I guess we need to make sure that we upgrade the client / server at the same time or only add new enum members at the end?

Reply:
There is probably a evolution of the protocol that would break this somehow, but yeah, in general adding new enum members at the end and simultaneous client / server upgrades should resolve it.

Reply:
> and simultaneous client / server upgrades should resolve it.

Always sketchy. :) We have version protocols within pgd. Perhaps we should think about that here too. (Not necessarily this second)

Reply:
Right, but in this case both the server (seafowl) and the client (seafowl_pg) are part of the same distribution and will be upgraded in lockstep, as opposed to different nodes in the cluster.